src/utils/cosmosdb_model.ts

```diff
@@ -30,6 +29,30 @@ export const BaseModel = t.interface({
 
 export type BaseModel = t.TypeOf<typeof BaseModel>;
 
+/**
+ * Model a tuple which defines the references to search for a document.
+ * A Cosmodb document must be looked-up by its Identity alongside its PartitionKey. If PartitionKey field is the same of Identity field, it can be omitted.
+ * Hence this type models both cases: (ID) or (ID,PK) respectively if PK literal type is omitted or provided.
+ * @param T the type of the document mapped by the model
+ * @param ID the literal type defining the ID field for the document.
+ * @param PK (optional) the literal type defining the eventual partition key for the document. Default: undefined
+ */
+export type DocumentSearchKey<
+  // T might not include base model fields ("id"), but we know they are mandatory in cosmos documents
+  // Quick win would be to narrow T to extend BaseModel, but that way we'd lose usage flexibility
+  // Hence we omit "extends BaseModel", but we check keys to be part of "T & BaseModel"
+  T,
+  ID extends keyof (T & BaseModel),
+  PK extends undefined | keyof (T & BaseModel) = undefined
+> = (T & BaseModel)[ID] extends string | number // narrow type to the ones that might be an identity
```

Comment:
in which case pk may be a number? (the literal type of pk field name)

Reply:
I believe the PK is always a `string`

Reply:
true